### PR TITLE
Fix Narayana recovery manager's too early start

### PR DESCRIPTION
**Pull request by gytis**
A fix to make sure that Narayana recovery manager doesn't start until application is ready.

### CODE REVIEW

**Comment by gytis** [comment]
CI failure seems to be unrelated.

**Comment by philwebb** [comment]
Thanks, do have a feeling for how much of a problem this currently causes? I'm trying to work out if we should apply this to the 1.5.x branch or if it can be deferred until 2.0.

**Comment by gytis** [comment]
@philwebb I think it should be applied for both master and 1.5.x. I can raise a separate PR if that would make it easier.

File: spring-boot/src/main/java/org/springframework/boot/jta/narayana/NarayanaRecoveryManagerBean.java [hunk 1]
```diff
@@ -41,8 +40,8 @@ public NarayanaRecoveryManagerBean(RecoveryManagerService recoveryManagerService
 		this.recoveryManagerService = recoveryManagerService;
 	}
 
-	@Override
-	public void afterPropertiesSet() throws Exception {
+	@EventListener
+	public void create(ApplicationReadyEvent ignored) throws Exception {
```
Comment (wilkinsona) [inline spring-boot/src/main/java/org/springframework/boot/jta/narayana/NarayanaRecoveryManagerBean.java]: Moving this to application ready makes me a little nervous. What is the implication for any transactional logic that tries to run earlier? That could even be in another listener to the `ApplicationReadyEvent` that happens to run earlier than this one.

File: spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/transaction/jta/NarayanaJtaConfiguration.java [hunk 2]
```diff
@@ -114,10 +115,12 @@ public TransactionManager narayanaTransactionManager() {
 	@Bean
 	@DependsOn("narayanaConfiguration")
 	public RecoveryManagerService narayanaRecoveryManagerService() {
+		RecoveryManager.delayRecoveryManagerThread();
```
Comment (wilkinsona) [inline spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/transaction/jta/NarayanaJtaConfiguration.java]: Given that this is a static method, what are the implications for integration tests? Will this still work if transaction recovery runs multiple times in the same JVM?

File: spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/transaction/jta/NarayanaJtaConfiguration.java [hunk 3]
```diff
@@ -114,10 +115,12 @@ public TransactionManager narayanaTransactionManager() {
...
 		return new RecoveryManagerService();
 	}
 
 	@Bean
+	@ConditionalOnMissingBean
```
Comment (wilkinsona) [inline spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/transaction/jta/NarayanaJtaConfiguration.java]: Can you please explain how this relates to delaying  starting the recovery manager?

File: spring-boot/src/main/java/org/springframework/boot/jta/narayana/NarayanaRecoveryManagerBean.java [hunk 4]
```diff
@@ -53,20 +52,17 @@ public void destroy() throws Exception {
...
+		getXARecoveryModule().addXAResourceRecoveryHelper(xaResourceRecoveryHelper);
 	}
 
-	private XARecoveryModule getXARecoveryModule(RecoveryManager recoveryManager) {
-		for (RecoveryModule recoveryModule : recoveryManager.getModules()) {
-			if (recoveryModule instanceof XARecoveryModule) {
-				return (XARecoveryModule) recoveryModule;
-			}
+	private XARecoveryModule getXARecoveryModule() {
+		XARecoveryModule xaRecoveryModule = XARecoveryModule.getRegisteredXARecoveryModule();
```
Comment (wilkinsona) [inline spring-boot/src/main/java/org/springframework/boot/jta/narayana/NarayanaRecoveryManagerBean.java]: Given that this is a static method, what are the implications for integration tests? Will this still work if transaction recovery runs multiple times in the same JVM?

File: spring-boot/src/main/java/org/springframework/boot/jta/narayana/NarayanaRecoveryManagerBean.java (hunk 1 above)
Comment (gytis) [inline spring-boot/src/main/java/org/springframework/boot/jta/narayana/NarayanaRecoveryManagerBean.java]: Why would there be any transaction logic happening before the application has started?

**Comment by gytis** [comment]
@wilkinsona I agree on decreasing the scope that Spring Boot needs to know about Narayana. I've pinged you on Gitter to discuss that.

File: spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/transaction/jta/NarayanaJtaConfiguration.java (hunk 3 above)
Comment (gytis) [inline spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/transaction/jta/NarayanaJtaConfiguration.java]: You're correct, that doesn't relate with delaying. However, it does allow application to override recovery bean to have more control. Should I move this annotation to a separate commit?

File: spring-boot/src/main/java/org/springframework/boot/jta/narayana/NarayanaRecoveryManagerBean.java (hunk 1 above)
Comment (wilkinsona) [inline spring-boot/src/main/java/org/springframework/boot/jta/narayana/NarayanaRecoveryManagerBean.java]: A user is free to do pretty much whatever they want during application startup. Beyond and application's own logic, there's also technologies like Flyway, Liquibase, and Hibernate to consider. I don't think Flyway or Liquibase will ever use JTA transactions but Hibernate definitely can. Hibernate will access the database during startup and I believe it may use transactions when it does so. Spring Data JPA also makes calls to Hibernate during its initialisation. I"m not sure if those calls are transactional or not but it needs to be considered here.

File: spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/transaction/jta/NarayanaJtaConfiguration.java (hunk 3 above)
Comment (wilkinsona) [inline spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/transaction/jta/NarayanaJtaConfiguration.java]: Yes please.

File: spring-boot/src/main/java/org/springframework/boot/jta/narayana/NarayanaRecoveryManagerBean.java (hunk 4 above)
Comment (gytis) [inline spring-boot/src/main/java/org/springframework/boot/jta/narayana/NarayanaRecoveryManagerBean.java]: It is just a getter which doesn't change any state. Internal logic is the same as the one I'm replacing here.

File: spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/transaction/jta/NarayanaJtaConfiguration.java (hunk 2 above)
Comment (gytis) [inline spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/transaction/jta/NarayanaJtaConfiguration.java]: This method sets a boolean to `true`. It makes sure that recovery manager has to be started explicitly. Without it recovery manager is started automatically when any transaction logic is executed for the first time. This is one of the candidates to be moved out as per our discussion. But for now it makes sense here.
As for integration tests, I'm not aware of any. Do you have any integration tests which I should check?

**Comment by gytis** [comment]
@wilkinsona I've updated the PR based on your comments

File: spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/transaction/jta/NarayanaJtaConfiguration.java (hunk 2 above)
Comment (wilkinsona) [inline spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/transaction/jta/NarayanaJtaConfiguration.java]: I was thinking of a user's integration tests for an application using Narayana where multiple application context's may be started and stopped. From your description of what the delay does, it sounds like that won't be a problem though.

**Comment by wilkinsona** [comment]
@gytis Have we lost some changes? All I can see now is addition of the call to delay the recovery manager. IIRC, previously there were also some calls to explicitly start the recovery manager. Are they no longer necessary for some reason?

File: spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/transaction/jta/NarayanaJtaConfiguration.java (hunk 2 above)
Comment (wilkinsona) [inline spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/transaction/jta/NarayanaJtaConfiguration.java]: > Without it recovery manager is started automatically when any transaction logic is executed for the first time

Wouldn't that be a good thing? It would address my earlier concern about recovery perhaps being performed too late. But I'm confused about that now as those changes seem to have disappeared.

**Comment by gytis** [comment]
@wilkinsona recovery manager is still being started explicitly as before (https://github.com/spring-projects/spring-boot/blob/master/spring-boot/src/main/java/org/springframework/boot/jta/narayana/NarayanaRecoveryManagerBean.java#L46). I've just reverted the change to start it when application is ready.
I've also extracted a couple of changes to the separate pull requests to keep things cleaner.

**Comment by wilkinsona** [comment]
Thanks. What versions of Narayana is this applicable to? Should the change go in 1.5 or only in master?

**Comment by gytis** [comment]
@wilkinsona master and 1.5.x would be great.
